### PR TITLE
Replace TypeName modifier methods with copy()

### PR DESCRIPTION
**Pull request by Egorand**

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/CodeWriter.kt [hunk 1]
```diff
@@ -332,7 +332,7 @@ internal class CodeWriter constructor(
       nameResolved = resolved != null
 
       // We don't care about nullability and type annotations here, as it's irrelevant for imports.
-      if (resolved == c.asNonNull().withoutAnnotations()) {
+      if (resolved == c.copy(nullable = false, annotations = emptyList())) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/CodeWriter.kt]: neat

File: src/main/java/com/squareup/kotlinpoet/TypeName.kt [hunk 2]
```diff
@@ -64,9 +64,12 @@ import kotlin.reflect.KVariance
  * Create new reference types like `com.example.HelloWorld` with [ClassName.bestGuess]. To build composite
  * types like `Set<Long>`, use the factory methods on [ParameterizedTypeName], [TypeVariableName],
  * and [WildcardTypeName].
+ *
+ * @param S the self-type, allowing [copy] to return the subclass type
  */
-sealed class TypeName constructor(
-  val nullable: Boolean, annotations: List<AnnotationSpec>
+sealed class TypeName<out S : TypeName<S>> constructor(
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/TypeName.kt]: Ick. I don’t think the cost is worth the benefit. Would rather we just override all the corresponding methods to narrow the return type

Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/TypeName.kt]: Agree, it's super noisy.. Will work on v2

File: src/test/java/com/squareup/kotlinpoet/AnnotatedTypeNameTest.kt [hunk 3]
```diff
@@ -33,52 +33,52 @@ class AnnotatedTypeNameTest {
...
   }
 
   @Test fun annotatedTwice() {
     val expected = "@$NN @java.lang.Override kotlin.String"
     val type = String::class.asTypeName()
-    val actual = type.annotated(NEVER_NULL)
-        .annotated(AnnotationSpec.builder(Override::class).build())
+    val actual = type
+        .copy(annotations = type.annotations + NEVER_NULL +
+            AnnotationSpec.builder(Override::class).build())
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/AnnotatedTypeNameTest.kt]: for later, should we have something like `fun Class<out Annotation>.asAnnotationSpec()` ?

Comment (Egorand) [reply src/test/java/com/squareup/kotlinpoet/AnnotatedTypeNameTest.kt]: Yeah, we should

**Pull request merged**